### PR TITLE
Add nfo thumb tag support

### PR DESCRIPTION
**Pull request by Ullmie02**
**Changes**
Add support for nfo thumb tag including tests.

- Only the first thumb tag for each artwork type is used because only one image for each artwork type is supported.
- Artwork defined in the nfo file takes precedence over local images in the media folder and provider images.

I would like to document this somewhere but don't know where the right place would be.

**Issues**
`<thumb>` part of #5262

### CODE REVIEW

**Review by crobibero — APPROVED** [review]

File: MediaBrowser.XbmcMetadata/Parsers/BaseNfoParser.cs [hunk 1]
```diff
@@ -785,6 +789,78 @@ protected virtual void FetchDataFromXmlNode(XmlReader reader, MetadataResult<T>
                         break;
                     }
 
+                case "thumb":
+                    {
```
Comment (barronpm) [inline MediaBrowser.XbmcMetadata/Parsers/BaseNfoParser.cs:793]: Can be done in a later PR, but I don't like this switch statement being so long. It should probably be changed either to a polymorphic solution or just method calls

**Review by barronpm — APPROVED** [review]

File: MediaBrowser.XbmcMetadata/Parsers/BaseNfoParser.cs [hunk 2]
```diff
@@ -785,6 +789,78 @@ protected virtual void FetchDataFromXmlNode(XmlReader reader, MetadataResult<T>
...
+                            "landscape" => ImageType.Thumb,
+                            "clearart" => ImageType.Art,
+                            // unknown type (including "poster") --> primary
+                            _ => ImageType.Primary,
+                        };
+
+                        Uri uri;
+                        try
+                        {
+                            uri = new Uri(val);
```
Comment (Bond-009) [inline MediaBrowser.XbmcMetadata/Parsers/BaseNfoParser.cs:820]: Use `Uri.TryCreate` instead

File: MediaBrowser.XbmcMetadata/Parsers/BaseNfoParser.cs [hunk 3]
```diff
@@ -785,6 +789,78 @@ protected virtual void FetchDataFromXmlNode(XmlReader reader, MetadataResult<T>
...
+                        // skip:
+                        // - empty aspect tag
+                        // - empty uri
+                        // - tag containing '.' because we can't set images for seasons, episodes or movie sets within series or movies
+                        if (string.IsNullOrEmpty(artType) || string.IsNullOrEmpty(val) || artType.Contains('.', StringComparison.Ordinal))
+                        {
+                            break;
+                        }
+
+                        ImageType imageType = artType switch
```
Comment (Bond-009) [inline MediaBrowser.XbmcMetadata/Parsers/BaseNfoParser.cs:806]: Maybe worthy of a helper method?

**Review by Bond-009 — CHANGES_REQUESTED** [review]

File: MediaBrowser.XbmcMetadata/Parsers/BaseNfoParser.cs [hunk 4]
```diff
@@ -803,25 +803,11 @@ protected virtual void FetchDataFromXmlNode(XmlReader reader, MetadataResult<T>
...
-
-                        Uri uri;
-                        try
-                        {
-                            uri = new Uri(val);
-                        }
-                        catch (UriFormatException ex)
+                        ImageType imageType = GetImageType(artType);
+
+                        if (!Uri.TryCreate(val, UriKind.Absolute, out var uri) || uri == null)
```
Comment (Bond-009) [inline MediaBrowser.XbmcMetadata/Parsers/BaseNfoParser.cs:808]:
```suggestion
                        if (!Uri.TryCreate(val, UriKind.Absolute, out var uri))
```
Can't be null when it returns true

**Review by Bond-009 — CHANGES_REQUESTED** [review]

**Comment by Ullmie02** [comment]
GitHub mobile is stupid. Sorry for removing the reviews

File: MediaBrowser.XbmcMetadata/Parsers/BaseNfoParser.cs (hunk 1 above)
Comment (Ullmie02) [reply MediaBrowser.XbmcMetadata/Parsers/BaseNfoParser.cs:793]: I have started working on a PR to use methods and make the switch statement shorter

File: tests/Jellyfin.XbmcMetadata.Tests/Parsers/MovieNfoParserTests.cs [hunk 5]
```diff
@@ -52,8 +55,36 @@ public MovieNfoParserTests()
...
+            else
+            {
+                _localImageFileMetadata = new FileSystemMetadata()
+                {
+                    Exists = true,
+                    FullName = "C:\\media\\movies\\Justice League (2017).jpg"
+                };
+                directoryService.Setup(x => x.GetFile(_localImageFileMetadata.FullName))
+                    .Returns(_localImageFileMetadata);
+            }
```
Comment (Bond-009) [inline tests/Jellyfin.XbmcMetadata.Tests/Parsers/MovieNfoParserTests.cs:78]:
```suggestion
                _localImageFileMetadata = new FileSystemMetadata()
                {
                    Exists = true,
                    FullName = MediaBrowser.Common.System.OperatingSystem.Id == OperatingSystemId.Windows ? "C:\\media\\movies\\Justice League (2017).jpg" : "/media/movies/Justice League (2017).jpg";
                };
                directoryService.Setup(x => x.GetFile(_localImageFileMetadata.FullName))
                    .Returns(_localImageFileMetadata);
```

**Review by Bond-009 — CHANGES_REQUESTED** [review]

File: MediaBrowser.Providers/Manager/MetadataService.cs [hunk 6]
```diff
@@ -706,6 +706,12 @@ public bool CanRefreshPrimary(Type type)
 
                     if (localItem.HasMetadata)
                     {
+                        foreach (var remoteImage in localItem.RemoteImages)
+                        {
+                            await ProviderManager.SaveImage(item, remoteImage.url, remoteImage.type, null, cancellationToken).ConfigureAwait(false);
+                            refreshResult.UpdateType = refreshResult.UpdateType | ItemUpdateType.ImageUpdate;
```
Comment (Bond-009) [inline MediaBrowser.Providers/Manager/MetadataService.cs:712]:
```suggestion
                            refreshResult.UpdateType |= ItemUpdateType.ImageUpdate;
```

**Review by Bond-009 — APPROVED** [review]

File: MediaBrowser.Providers/Manager/MetadataService.cs (hunk 6 above)
Comment (Ullmie02) [reply MediaBrowser.Providers/Manager/MetadataService.cs]: There were more of lines like this which I've changed as well. Hope it's okay.

**Comment by olympus123** [comment]
I tried it and it works, it imports poster, banner, clearart, clearlogo, discart and landscape.
But it doesn't seem to import fanart. Does it support fanart?
I tried

  <fanart>
    <thumb>https://assets.fanart.tv/fanart/movies/141052/moviebackground/justice-league-5793f518c6d6e.jpg</thumb>
  </fanart>

  <backdrop>
    <thumb>https://assets.fanart.tv/fanart/movies/141052/moviebackground/justice-league-5793f518c6d6e.jpg</thumb>
  </backdrop>

<thumb aspect="fanart">https://assets.fanart.tv/fanart/movies/141052/moviebackground/justice-league-5793f518c6d6e.jpg</thumb>

<thumb aspect="backdrop">https://assets.fanart.tv/fanart/movies/141052/moviebackground/justice-league-5793f518c6d6e.jpg</thumb>

**Comment by olympus123** [comment]
Thank you so much for this, I can finally put the artwork outside of my media or even create nfo files pointing to http links for artwork.
I tried it and it works for poster, banner, landscape, clearlogo, clearart and discart.
But it doesn't seem to support fanart.
I tried

`  <thumb aspect="fanart">https://assets.fanart.tv/fanart/movies/141052/moviebackground/justice-league-5793f518c6d6e.jpg</thumb>`

`<thumb aspect="backdrop">https://assets.fanart.tv/fanart/movies/141052/moviebackground/justice-league-5793f518c6d6e.jpg</thumb>`

I tried also with the fanart tag

`  <fanart>
    <thumb>https://assets.fanart.tv/fanart/movies/141052/moviebackground/justice-league-5793f518c6d6e.jpg</thumb>
  </fanart>`

**Comment by Ullmie02** [comment]
@olympus123 I'll add support for the fanart tag with a later PR

**Comment by olympus123** [comment]
Thank you so much for all the work you do for nfo parsing in Jellyfin.
I know it's not the right place to ask you, but I can't resist.
There is something about nfo parsing that I really like in Kodi.
Kodi and Jellyfin both support strm files. When you scrape them both can't "understand" the file and add to the library the info of the actual file, I mean information about the video stream, the audio stream etc, the mediainfo of the file.
In Kodi if you create an nfo for the strm file and add the fileinfo / streamdetails in it yourself, Kodi will read them from the nfo and import the mediainfo details of the file.
I would really love if there was a way for Jellyfin to do the same.
I mean "nfo fileinfo tag support".

**Comment by olympus123** [comment]
> @olympus123 I'll add support for the fanart tag with a later PR

Hi, I have just tested latest Kodi.
They have enhanced the nfo thumb tag a lot.
The nfo thumb tag supports fanart too, no need  for a seperate fanart tag, Kodi 19 can read the fanart aspect from the thumb tag now.
This is what I am talking about, it works in Kodi 19 to cover fanart, a sererate fanart tag is not needed.

<thumb aspect="fanart">https://assets.fanart.tv/fanart/movies/141052/moviebackground/justice-league-5793f518c6d6e.jpg</thumb>

Also Kodi can now hadle the collections  artwork too with the movie's nfo thumb tag.

  <thumb aspect="set.poster">blabla</thumb>
  <thumb aspect="set.landscape">blabla</thumb>
  <thumb aspect="set.banner">blabla</thumb>
  <thumb aspect="set.clearlogo">blabla</thumb>
  <thumb aspect="set.clearart">blabla</thumb>
  <thumb aspect="set.fanart">blabla</thumb>

**Comment by Ullmie02** [comment]
> `<thumb aspect="fanart">https://assets.fanart.tv/fanart/movies/141052/moviebackground/justice-league-5793f518c6d6e.jpg</thumb>`

could be added fairly easily. Thanks for the heads up!

Unfortunately, the collection artwork is not that easy. Collections are completely different entities in our database and I haven't found a way yet to set pictures for them with the movie nfo files

**Pull request merged**